i18n: Write language name in the language itself.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is a small refactor.
- Second commit is to write language name in the language itself in language dropdown and modal.

Fixes: <!-- Issue link, or clear description.-->#14565

https://chat.zulip.org/#narrow/channel/58-translation/topic/writing.20language.20name.20in.20the.20language.20itself/with/2305636

**How changes were tested:**

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
| Before | After |
| ------ | ------ |
| <img width="1365" height="766" alt="image" src="https://github.com/user-attachments/assets/a584db08-d928-4644-be82-e1a58de38de4" /> | <img width="1365" height="766" alt="image" src="https://github.com/user-attachments/assets/227e63f0-5c65-4423-b488-c0d62b56fc94" /> |
| [Screencast from 2025-11-20 18-56-42.webm](https://github.com/user-attachments/assets/1a199b5a-0475-4f36-96ab-6fec5949ff76) | [Screencast from 2025-11-20 18-57-24.webm](https://github.com/user-attachments/assets/3a66d59c-2292-40cf-9109-831d726cde04) |



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/i18n.ts:
```diff
@@ -60,12 +60,29 @@ export function $t_html(
+function get_language_display_name(language_code: string): string {
+    const locale = new Intl.Locale(language_code === "en" ? "en-US" : language_code);
+    let language_display_name = new Intl.DisplayNames([locale], {type: "language"})
+        .of(locale.language)!
```
Have used `!` here assuming the server would send language codes that will have a display name value. Can add a condition to show the original name if `Intl.DisplayNames` returns undefined.

timabbott replied:
Probably fine to do it this way, failing visibly seems better here.

sahil839 commented:
Added code to hard-code the display name for two languages as discussed [on CZO](https://chat.zulip.org/#narrow/channel/58-translation/topic/writing.20language.20name.20in.20the.20language.20itself) and also updated the ordering of languages.

<img width="956" height="674" alt="image" src="https://github.com/user-attachments/assets/c59639de-a79d-4ab5-a7db-c961cfa446b1" />

alya commented:
I'm a bit confused -- that screenshot looks like the old UI, not what's currently in `main`?

timabbott commented on web/src/settings_components.ts:
```diff
@@ -2024,7 +2024,9 @@ export function get_channel_folder_value_from_dropdown_widget($elem: JQuery): nu
 export const language_options = (): Option[] => {
-    const languages = get_language_list_columns(realm.realm_default_language);
+    const languages = get_language_list_columns(realm.realm_default_language).toSorted((a, b) =>
+        util.strcmp(a.name.toLowerCase(), b.name.toLowerCase()),
```
Empirically it seems not to, though, marking as resolved.

timabbott commented:
Merged, thanks @sahil839 and @andersk!
